File: Test_2/Strings/Multiply_Strings/Multiply_Strings.java
// Problem Link -> https://leetcode.com/problems/multiply-strings/description/

class Solution {
    public String multiply(String num1, String num2) {
        StringBuilder ref_num1 = new StringBuilder(num1);
        StringBuilder ref_num2 = new StringBuilder(num2);
        if(num1.equals("0") || num2.equals("0")) return "0";     
        
        ref_num1 = ref_num1.reverse();
        ref_num2 = ref_num2.reverse();

        ArrayList<Integer> v = new ArrayList<>(Collections.nCopies(ref_num1.length() + ref_num2.length(), 0));

        for (int i = 0; i < ref_num1.length(); i++) {
            int digit1 = Character.getNumericValue(ref_num1.charAt(i));
            for (int j = 0; j < ref_num2.length(); j++) {
                int digit2 = Character.getNumericValue(ref_num2.charAt(j));
                int currIndex = i + j;
                int currProduct = digit1 * digit2 + v.get(currIndex);
                int currVal = currProduct % 10;
                int currCarry = currProduct / 10;
                v.set(currIndex, currVal);
                v.set(currIndex + 1, v.get(currIndex + 1) + currCarry);
            }
        }

        Collections.reverse(v);
        StringBuilder res = new StringBuilder();
        boolean notLeadingZeros = false;
        for (int i = 0; i < v.size(); i++) {
            if(!notLeadingZeros && v.get(i) == 0) continue;
            if(v.get(i) != 0 && notLeadingZeros == false) {
                notLeadingZeros = true;
            }
            res.append(v.get(i));
        }
        return res.toString();
    }
}

File: Test_2/Strings/Multiply_Strings/Multiply_Strings.cpp
// Problem Link -> https://leetcode.com/problems/multiply-strings/description/

class Solution {
public:
    string multiply(string num1, string num2) {
        if(num1 == "0" || num2 == "0") return "0";
        vector<int>v(num1.size()+num2.size(), 0);
        
        reverse(num1.begin(),num1.end());
        reverse(num2.begin(),num2.end());

        for(int i = 0; i < num1.size(); i++) {
            int digit1 = num1[i]-'0';
            for(int j = 0; j < num2.size(); j++) {
                int digit2 = num2[j]-'0';
                int currIndex = i + j;
                int currProduct = digit1 * digit2 + v[currIndex];
                int currVal = currProduct%10;
                int currCarry = currProduct/10;
                v[currIndex+1] += currCarry;
                v[currIndex] = currVal;
            }
        }
        string res = "";
        reverse(v.begin(),v.end());
        bool flag = false;
        for(auto it : v) {
            if(!flag && it == 0) continue;
            if(it != 0) flag = true;
            res += to_string(it);
            
        }
        return res;
    }
};